rfcs/text/0014_scheduled_reports_new_plugin.md

```diff
@@ -0,0 +1,57 @@
+- Reporting job's themselves have some implementation concerns embedded into them with regards to queueing, making them less portable.
+
+Writing scheduled reports as a new plugin can also provide us with an opportunity to fix long standing technical issues and ownership boundaries.
+
+- Our PDF generation process requires numerous snapshots of the page, and generating the PDF in memory, versus using native headless PDF methods.
+- A lot of logic is embedded in the reporting stack that belongs elsewhere. [Prior RFCs were established for this here](https://github.com/elastic/kibana/pull/64372), and this new plugin provides us an opportunity to work on this as well.
+- Generating larger reports has been a historic challenge in cloud environments, and a new plugin would allow us to explore more streamlined methods without risking a breakages in our current plugin.
+
+At it's core, our motivation is to get this feature shipped without further delay. Prior projects that were similar in level's of effort have often been delivered near (or after) internal deadlines. Our movement to Kibana's new platform is a good example.
+
+# Overall design
+
+The new plugin would follow Kibana's recommended and best practices, use API Keys internally, as well as TaskManager for orchestrating and running of reporting jobs. At creation, reporting would:
+
+1. Validate the request, license, and user permission levels -- making use of the new platform features.
+2. If a request is indeed valid, a Task is registered into Task Manager, and some meta data pertaining to the job is recorded in our own index.
+3. At job execution time, reporting would use the metadata from the job to execute the work required. [More details can be found here](https://github.com/elastic/kibana/pull/64372).
+    1. Any logic needed to render properly should be embedded into the page being reported.
+    2. Once rendering is complete in our headless browser, a PDF or PNG of the page is captured.
+    3. At no point should our reporting plugin "inject" custom CSS or JS in order to export the page.
+4. Once complete, the report is captured into ElasticSearch.
+5. The report is then exported via the connector chosen in Step 1.
+
+After some period of time, gauging stability and overall hardiness, we should endeavor to migrate the older reporting plugin's behavior to the fundamentals in the newly created one. Our intention isn't to create and manage two separate plugins, but create a plugin that can support all use cases without having to make massive changes to our current plugin.
+
+Of course this strategy isn't without its shortcoming.
+
+# Drawbacks
+
+While authoring a new plugin allows us a chance to create from the ground-up, there's significant potentials for failure.
+
+- The current plugin has many historical issues fixed and implemented, and writing a new plugin means we're potentially susceptible to those.
```

Comment:
Are there parts of the existing implementation that are roughly in the shape we'd want them moving forward? If so, can they be carried over to the new plugin implementation?  Doing that may help mitigate some of this risk.

Reply:
++, I envision that we'd have some kind of session regarding what can stay/go and how we want to move forward with it.

Reply:
Yes, exactly: there's actually a fair bit of code that I'd recommend copying over:

- The browser-driver factory and all of it's related pieces.
- All of the Reporting-storage code, which handles the saving of the final "rasterized" asset.
- The interfaces behind our export-registry, which handles the gathering aspects of what type of assets Kibana can report on.

Some of these modules to have some code that handles things like es-queue's `cancellation-token`, which we'll want to kill or adapt since it's not an idiom that Task manager likely understands.